README.md
# DataCAD_simple_math_core

Plugin for [complex rest](https://github.com/ISGNeuroTeam/complex_rest/tree/develop)

## Getting Started

These instructions will get you a copy of the project up and running on your local machine for development and testing purposes. See deployment for notes on how to deploy the project on a live system.

### Prerequisites

Deploy [complex rest](https://github.com/ISGNeuroTeam/complex_rest/tree/develop)

### Installing

* Make symlink for ./DataCAD_simple_math_core/DataCAD_simple_math_core in plugins directory
* Run complex rest server

## Running the tests
Run all tests:
```bash
python ./complex_rest/manage.py test ./plugin_dev/dtcd_simple_math_core/tests --settings=core.settings.test
```

## Deployment

* Make plugin archive:
```bash
make pack
```
* Unpack archive into complex_rest plugins directory

## Built With

* [Django](https://docs.djangoproject.com/en/3.2/) - The web framework used


## Contributing

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/your/project/tags). 

## Authors
Starchenkov Andrey
Serditov Nikita

## License

[OT.PLATFORM. License agreement.](LICENSE.md)

## Features
- Creates swt table when graph is calculated first time and has no swt table.
- Supports importing data through inPort/outPort in terms of one graph. As it works with data from swt table, not graph, then sometimes it requires two consequent graph calculations to calc all imported data.
- Support `SWT` and `SWT Export` data import from a different graph only if source graph has the same `_t` tick value as the latest `-t` tick value of the target swt table. Works with "Загрузка данных из соседних вкладок" custom action only.
- Support data import from data lake node. Works with "Загрузка данных из озера данных" custom action only.

## Setup of the plugin
Go to dtcd_simple_math_core/dtcd_simple_math_core.conf and edit it:
- set logging `level` at the [logging] section.
- set parameters of [ot_simple_connector], usually `host` is `localhost`, all the rest parameters get from your administrator.
- set `swt_line_index` parameter to `PREVIOUS_MONTH` if you need to fill graph with fact swt table values, or `LATEST` if you need to fill graph with random latest row of swt table.
- set `time_offset` parameter to show the difference between timezone used to set wide swt table `_t` column values and utc timezone. For example if `_t` values were generated in `utc+3` timezone, the `time_offset` must be set to `3`.

CHANGELOG.md
## [0.1.9] - 2023-08-22

### Added 
- `time_offset` parameter to establish the difference between timezone where `_t` column values of the `wide` swt table were generated and `utc` timezone.

### Changed
- now `PREVIOUS_MONTH` leads to fill the graph with values of the row of swt table which `_t` value is previous to current time **considering timezone**. So if today is 2nd of August 2023, then values from row with `1688158800` (30th of June 2023 21:00:00) at `_t` column will be used.

## [0.1.8] - 2023-08-15

### Changed

- graph is filled with values of the different row of the swt table now, depending on `swt_line_index` config parameter.
  - `PREVIOUS_MONTH` leads to fill the graph with values of the row of swt table which `_t` value is previous to current time. So if today is 2nd of August 2023, then values from row with `1690837200` (31st of July 2023) at `_t` column will be used 
  - `LAST` leads to fill graph with values of the last row of current swt table. Which may be **not the latest**.     

## [0.1.7] - 2023-08-10

### Fixed

- GET swt request now returns **an array** of an exact line of the swt table with index of `tick`, given in request.

### Changed

- if `tick` sent not as `digit`, plugin returns an error message.
- if `tick` is not sent at all, default value of `-1` is used.

## [0.1.6] - 2023-08-09

### Fixed

- GET swt request now returns exact line of the swt table with index of `tick`, given in request
- if swt table happen to be sparse, smc now increases the density of it according to its schema

## [0.1.5] - 2023-08-04

### Fixed
- too many values to unpack error

## [0.1.4] - 2023-08-03

### Fixed
- added support of string enclosed with singular quotes used in `select` values. Example: regular string `"text"`, string from `select` entry: `"'text''"`. Never use singular quotes to give `string` values, always double. Singular quotes used in OTL to define columns only.

### Changed
- replaced import of data from external swt tables with simple swt calc

## [0.1.3] - 2023-07-27

### Changed

- update of the `ot_simple_connector` version to 0.1.12
- added `property_globals` section to dtcd_simple_math_core.conf.example. Please update your *.conf file.

### Fixed

- parsing inPort values when not surrounded by whitespaces. Example: `(inPort1 + inPort2 + inPort3) / 100` evaluated to `(inPort1 + 'DataNode_1.value' + inPort3) / 100`.
- parsing float values as float values, not strings. Example: `eval 'DataNode_1.value' = '0.6'` did not evaluate.

## [0.1.2] - 2023-07-20

### Changed

- `ot_simple_connector` version updated up to 0.1.11

### Fixed

- 400 error on a first calc of a graph with DataLakeNode in it

## [0.1.1] - 2023-07-18

### Changed

- `ot_simple_connector` version updated up to 0.1.10

## [0.1.0] - 2023-07-10

### Added

- Creation of the swt table on a first graph calculation if it does not exist.
- Parametrized `otl_create_fresh_swt` template and regexp strings at the config file.
- Documentation.
- Tests.
- Import data through `SWT` and `SWT Export` property request type without any custom actions or buttons, from
  properties of the nodes from a different graph if source graph has the same `_t` tick value as the latest tick of
  target graph.
- Import data through ports of the nodes.

### Fixed

- Cyrillic symbols usage problem.
- Parsing of the `email` string in the `expression` of the property of the noe fixed.

### Changed

- Graph calculation is only through the `POST` method to `/dtcd_simple_math_core/v1/graph/` url with `swt_name`
  and `graph` parameters required.
- SWT may be read only through the `GET` method to `/dtcd_simple_math_core/v1/swt/` url with `swt_name` parameter
  required.

dtcd_simple_math_core/translator/graph.py
# -*- coding: utf-8 -*-
"""This module describes logic of analyzing and merging graph data
"""
from _datetime import datetime, timezone, timedelta
import json
import logging
import re
from typing import Dict, Union, List

from ..settings import GRAPH_GLOBALS, plugin_name, EVAL_GLOBALS, UTC_OFFSET
from .swt import SourceWideTable

from .node import Node
from .edges import Edge


class Graph:
    """This class describes how Graph works:

    Args:
        :: log: local instance of plugin logger
        :: nodes: dictionary of all nodes, consist of its names and its Node instances
        :: edges: list of all edges, consists of all port connections of nodes in graph
        :: name: name of current Graph
        :: dictionary: raw json object to keep, update and send back on api request
        :: swt_imported_tables: list of strings that represent names of the swt_tables
                                that are required to be imported
    """
    log: logging.Logger = logging.getLogger(plugin_name)
    nodes: Dict[str, Node]
    edges: List[Edge]
    name: str
    dictionary: Dict
    swt_import_tables: List[str]

    def __init__(self, name: str, graph: Union[Dict, str]):
        self.nodes = {}
        self.edges = []
        self.name = name
        self.dictionary = graph
        self.swt_import_tables = []

    def initialize(self) -> None:
        """Here we parse the graph if it was given as a string.
        And then parse all the nodes of the graph from json to nodes dictionary in order
        to keep them in Nodes objects.
        """
        self.log.debug('started initialization of graph')
        self.dictionary = json.loads(self.dictionary) if isinstance(self.dictionary, str) \
            else self.dictionary
        self.parse_nodes()
        self.parse_edges()
        self.parse_ports_of_nodes()

    def parse_nodes(self) -> None:
        """We parse graph from json into Nodes objects
        """
        self.log.debug('started parsing the nodes...')
        for node in self.dictionary['graph']['nodes']:
            self.nodes[node['primitiveID']] = Node(node.get('primitiveID', ''))
            self.nodes[node['primitiveID']].initialize(node)
            self.log.debug("parsed node %s", node['primitiveID'])
            # self.swt_import_tables.extend(self.nodes[node['primitiveID']].swt_imported_tables)
        self.log.debug('parsed nodes successfully...')

    def parse_edges(self) -> None:
        """We parse graph from json to get name connections of inPorts and outPorts
        It will be a dict of pairs: {targetPort: sourcePort}
        """
        self.log.debug('started parsing the edges...')
        for edge in self.dictionary['graph']['edges']:
            self.edges.append(Edge(edge))
        self.log.debug('parsed edges successfully...')

    def parse_ports_of_nodes(self):
        """Function to parse ports of the nodes

        As node has OUT or IN ports, and they may be getting data from another nodes
        we must do this in graph layer.

        Here we loop through edges to know where we take data and where we should put it."""
        # pylint: disable=line-too-long
        for edge in self.edges:
            # get outPort value
            node = self.nodes[edge.source_node]
            source_port_expression = node.get_port_expression_by_primitive_id(edge.sp)
            if '.' not in source_port_expression:
                source_port_expression = '.'.join([edge.source_node, source_port_expression])
            if re.fullmatch(EVAL_GLOBALS['re_numbers'],
                            source_port_expression) is None and '.' in source_port_expression:
                source_port_expression = f"{source_port_expression}"
            # change inPort value
            self.nodes[edge.target_node].change_import_expression_by_primitive_id(edge.target_port,
                                                                                  source_port_expression)

    def filtered_columns(self, swr: List) -> List:
        """We get a row of the source wide table which is "Source Wide Row" >>> swr
        and filter out all strings that start with "_" symbol

        Args:
            swr: list of strings to filter

        Returns:
            list of filtered strings

        Example:
            input: swr = ['test', 'row', '_operations_order', '_size']
            output: ['test', 'row']
        """
        self.log.debug('started filtering columns at swr=%s', swr)
        result = list(filter(lambda c: not c.startswith("_"), swr))
        self.log.debug('result=%s', result)

        return result

    def get_property_of_the_node_by_id(self, object_id: str) -> Dict:
        """This function gets the actual property object out of saved raw json Graph
        by its 'primitiveID'

        It is required for its' further update

        Args:
            object_id: name of the object to find and return

        Return:
            dictionary with data of the object with name object_id found
            in raw json dictionary of the Graph

        TODO:
            * it seems that here we haven't checked the situation when object is not found
              this is actually impossible because we check if we have this object in graph
              before using this function.
        """
        self.log.debug('getting the properties object of the %s node', object_id)
        nodes = self.dictionary['graph']['nodes']
        for node in nodes:
            if node['primitiveID'] == object_id:
                self.log.debug('properties object is found')
                return node['properties']

        self.log.debug('%s node is not found, returning empty dictionary...', object_id)
        return {}

    def update_property_at_graph(self, node_name: str, prop_name: str, parameter: str, value: Union[str, dict]) -> None:
        """Function to update given property with data

        Args:
            node_name: name of the node where to look for property
            prop_name: name of the property to update
            parameter: name of the parameter of the property to update
            value: actual value to set to property
        """
        self.log.debug('updating prop_name=%s of the %s node with value=%s...',
                       prop_name, node_name, value)
        properties = self.get_property_of_the_node_by_id(object_id=node_name)
        if prop_name in properties:
            if parameter in properties[prop_name].keys():
                properties[prop_name][parameter] = value
            else:
                properties[prop_name].update({parameter: value})
        else:
            properties[prop_name] = {parameter: value}
        self.log.debug('update successfully done...')

    def update(self, swr: List) -> Dict:
        """Function to update nodes objects and actual raw json dictionary

        Args:
            swr: the least row of the source wide table
        """
        # updating graph with data from otl
        self.log.debug('updating the graph...')
        self.log.debug('input: swr=%s', swr)
        for column in self.filtered_columns(swr=swr):
            # here we need to check if columns returned from calculations are present in graph
            if not self.graph_has_this(column):
                continue
            reg_exp = GRAPH_GLOBALS['re_object_id_and_property']
            object_id, object_property = re.match(reg_exp, column).groups()

            self.log.debug('from column=%s we have object_id=%s and object_property=%s',
                           column, object_id, object_property)

            try:
                self.nodes[object_id].update_property(object_property, swr[column])
                self.log.debug('updated property at the self.nodes dictionary')
                self.update_property_at_graph(node_name=object_id, prop_name=object_property,
                                              parameter='value', value=swr[column])
                self.log.debug('updated property at the self.graph dictionary')

            except KeyError:
                self.log.debug('No %s node found, only %s got', object_id, self.nodes.keys())

        # updating graph with data from self.nodes, especially _operations_order property
        for node in self.nodes:
            if '_operations_order' in self.get_properties_of_the_node_from_dictionary_by_id(node_name=node):
                continue
            self.update_property_at_graph(node_name=node, prop_name='_operations_order',
                                          parameter='expression',
                                          value=self.nodes[node].properties[
                                              '_operations_order'].get_expression)
            self.update_property_at_graph(node_name=node, prop_name='_operations_order',
                                          parameter='type',
                                          value=self.nodes[node].properties['_operations_order'].type)
            self.update_property_at_graph(node_name=node, prop_name='_operations_order',
                                          parameter='value',
                                          value=self.nodes[node].properties['_operations_order'].value)
            self.update_property_at_graph(node_name=node, prop_name='_operations_order',
                                          parameter='title',
                                          value=self.nodes[node].properties['_operations_order'].__dict__.get(
                                              "title", ""))
            self.update_property_at_graph(node_name=node, prop_name='_operations_order',
                                          parameter='status',
                                          value=self.nodes[node].properties['_operations_order'].status)

        return self.dictionary

    def get_properties_of_the_node_from_dictionary_by_id(self, node_name: str) -> Dict:
        node_list = self.dictionary['graph']['nodes']
        for node in node_list:
            if node['primitiveID'] == node_name:
                return node['properties']

    def graph_has_this(self, column: str) -> bool:
        """function to check if column is in graph, because some swt imported data is saved
        from external swt tables and are not present at current graph, and must not.

        Args:
            ::column: name of the column to check, for example: "SWTNode_44.exportedProperty"

        Result:
            True if column is at the graph, False otherwise
        """
        if '.' not in column:
            return False

        columns = column.split('.')
        if len(columns) != 2:
            return False
        node_name, prop_name = tuple(columns)
        return node_name in self.nodes and prop_name in self.nodes[node_name].properties

    def calc(self) -> Dict:
        """Main calculation function of Graph
        """
        self.log.debug('calculating graph...')
        swt = SourceWideTable(self.name)
        swt.initialize()
        nodes_eval_expressions = self.get_nodes_eval_expressions()
        self.log.debug('nodes_eval_expressions: %s', nodes_eval_expressions)
        list_of_sw_rows = swt.calc(nodes_eval_expressions, self.swt_import_tables)
        self.log.debug('list_of_sw_rows[-1]=%s', list_of_sw_rows[-1])

        row_of_swt = get_row_of_swt(list_of_sw_rows, GRAPH_GLOBALS['swt_line_index'])
        result = self.update(row_of_swt)
        return result

    def get_nodes_eval_expressions(self) -> List[Dict]:
        """Function to get all the eval expressions for all nodes and properties
        """
        self.log.debug('getting nodes eval expressions...')
        sorted_nodes = self.get_sorted_nodes()
        eval_expressions = []
        for node in sorted_nodes:
            evals, imported_columns = node.get_eval_expressions()
            self.swt_import_tables.extend(imported_columns)
            eval_expressions.extend(evals)
        self.log.debug('eval_expressions=%s', eval_expressions)
        return eval_expressions

    def get_sorted_nodes(self) -> list[Node]:
        """Function to get the dictionary of all nodes sorted by their _operations_order
        and name

        Example:
            "nodes": [
                {
                    "primitiveID": "Data_685",
                    "primitiveName": "Data",
                    "properties": {
                        "type": {...},
                        "name": {...},
                        "description": {...},
                        "_operations_order": {        <<< this property is used
                            "status": "complete",
                            "type": "expression",
                            "expression": 100,        <<< this is what current property sorted by
                            "input": {
                                "component": "textarea"
                            }
        """
        self.log.debug('sorting the nodes by operations order...')
        try:
            result = sorted(self.nodes.values(),
                            key=lambda x: int(x.properties['_operations_order'].expression))
            self.log.debug('result=%s', result)
            return result

        except KeyError as error:
            raise Exception('Not all nodes have _operations_order property') from error

    def get_imported_data(self):
        ...

    def __str__(self) -> str:
        """Simple string representation of the Graph object
        """
        return f'{self.name=}\n' + '\n'.join(f'\t{node}' for node in self.nodes)

    @classmethod
    def read_from_file(cls, filename: str) -> 'Graph':
        """Function to read Graph from file on disk

        Args:
            filename: name of the file to read

        Returns:
            Function returns already parsed Graph object of the json read from file.
        """
        path = GRAPH_GLOBALS['path_to_graph'].format(filename)
        cls.log.debug('reading a graph %s at %s', filename, path)
        with open(path, encoding='utf-8') as file:
            return Graph(filename, file.read())


def get_row_of_swt(list_of_rows: List[Dict], swt_line_index: str) -> Dict:
    if swt_line_index == 'PREVIOUS_MONTH':
        threshold_time = get_threshold_time(offset=UTC_OFFSET)
        # Create a list comprehension that filters out dictionaries with `_t` value
        # larger than or equal to the current time
        filtered_list = [d for d in list_of_rows if int(d['_t']) < threshold_time]

        # Find the dictionary with the highest `_t` value in the filtered list
        result = max(filtered_list, key=lambda d: d['_t'])
    else:  # if swt_line_index is 'LAST' or anything else we return latest row
        result = list_of_rows[-1]

    return result


def get_threshold_time(offset: int) -> int:
    """ the problem with offset is that currently wide table is created with utc+03 timezone,
        so if it was required to have 01.08.2023 00:00:00, instead it stores 31.07.2023 21:00:00
        and all cells in wide swt table store datetime of 21:00:00 of the latest day of the month
        and threshold should be like this:

             required date        `threshold date`               today
        [30.06.2023 21:00:00] | [31.07.2023 21:00:00] | [22.08.2023 21:00:00] | [31.08.2023 21:00:00]

        if today is 22.08.2023 09:45:00.
        previous time cell  for current date is 31.07.2023 21:00:00
        next time cell      for current date is 31.08.2023 21:00:00
        then required data is from row with cell of 30.06.2023 21:00:00     <<< `threshold date`

        task of this function is to get threshold date out of today datetime using 'offset' parameter

        Args:
            ::offset: amount of hours offset used in wide swt table considering utc timezone

        Returns:
            integer value timestamp fof the `threshold date`
    """
    tz = timezone(timedelta(hours=offset))
    ct = datetime.now(tz=tz)
    threshold_time = datetime(year=ct.year, month=ct.month, day=1) - timedelta(hours=offset)
    return int(round(threshold_time.timestamp()))
